text/2023-04-26-color-outlines.md

```diff
@@ -0,0 +1,198 @@
+    glyph_id: GlyphId, 
+    palette_fn: impl Fn(u16) -> Option<Color>, 
+    pen: &mut impl ColorPen
+) -> Result<(), Error>
+```
+Loads a color outline for the specified glyph and invokes the appropriate
+callbacks on the pen.
+
+The `palette_fn` parameter is used to specify the mapping from palette
+color index to a color value. This can be used to implement overrides and
+a helper will also be provided for constructing this function for basic
+palette support.
+
+### The `ColorPen` trait
+
+This trait contains methods that are implemented by the caller to access
+the sequence of operations necessary to render a color outline.
+
+> Note: When provided below, the code in block quotes demonstrates how these
+operations might roughly be implemented for a Skia based renderer (assuming
+a local variable `canvas` of type `SkCanvas`)
+
+#### Methods
+
+```rust
+fn bounds(&mut self, bounds: BoundingBox)
+```
+This is guaranteed to be the first method called on the pen to allow an
+appropriately sized surface to be allocated if necessary. It will pass the
+value of the `ClipBox` if available, otherwise a bounding box computed
+during evaluation of the paint graph.
+
+In accordance with the spec, if the paint graph is determined to be
+unbounded, no methods on the pen will be called and the `color_outline`
+method will return an error.
+
+```rust
+fn push_transform(&mut self, transform: Transform)
+```
+Push a transform onto the state stack.
+
+> `canvas->save();`<br>
+`canvas->concat(TransformToSkMatrix(transform));`
+
+```rust
+fn pop_transform(&mut self)
+```
+Pop a transform from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_clip(&mut self, glyph_id: Glyph_id, path: &Path)
+```
+Push a clip path onto the state stack.
+
+> `SkPath clipPath = PathToSkPath(path);`<br>
+`canvas->save();`<br>
+`canvas->clip(clipPath, true);`
+
+```rust
+fn pop_clip(&mut self)
+```
+Pop a clip path from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_layer(&mut self, mode: Option<CompositeMode>)
+```
+Push a new layer to the state stack, optionally with a composite mode.
+
+If `mode` is not `None`:
+> `SkPaint blendModePaint;`<br>
+`blendModePaint.setBlendMode(CompositeModeToSkBlendMode(mode.unwrap());`<br>
+`canvas->saveLayer(nullptr, &blendModePaint);`
+
+Otherwise:
+> `canvas->saveLayer(nullptr, nullptr);`<br>
+
+```rust
+fn pop_layer(&mut self)
+```
+Pop a layer from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn fill(&mut self, brush: &Brush);
+```
+Fill the current clip with the given brush.
+
+> `canvas->drawPaint(BrushToSkPaint(brush));`<br>
+
+```rust
+fn fill_path(
+    &mut self,
+    glyph_id: GlyphId,
+    path: &Path,
+    brush: &Brush,
+    brush_transform: Option<Transform>
+)
+```
+Fill the specified path with the given brush. Note that this is an optimization
+for the pattern `Clip Transform* Fill` in the graph. The intermediate
+transforms are captured in the optional `brush_transform` parameter.
```

Comment:
Just FYI, the Skia version of this optimization does not support intermediate transforms, and would just drop the optimization and follow to the standard sequence of operations if there is a clip + transform + `Paint*Gradient`. So this latter parameter we might not need, or at least I haven't evaluated the performance benefit of that case.